File: src/main/java/gg/manny/forums/user/service/UserService.java
package gg.manny.forums.user.service;

import gg.manny.forums.rank.Rank;
import gg.manny.forums.rank.RankRepository;
import gg.manny.forums.user.User;
import gg.manny.forums.user.UserRepository;
import gg.manny.forums.user.grant.Grant;
import gg.manny.forums.util.MojangUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.*;

@Service
public class UserService implements UserDetailsService {

    @Autowired private UserRepository userRepository;
    @Autowired private RankRepository roleRepository;

    @Autowired private BCryptPasswordEncoder encoder;

    public User findUserByEmail(String email) {
        return userRepository.findByEmail(email);
    }

    public User findUserByName(String name) {
        return userRepository.findByUsername(name);
    }

    public void save(User user) {
        userRepository.save(user);
    }

    /**
     * Creating registered users for the first time
     *
     * @param user User to register
     */
    public void createUser(User user) throws Exception {
        UUID uniqueId = MojangUtils.fetchUUID(user.getUsername());
        if (uniqueId == null) throw new RuntimeException("Error finding uniqueId");

        user.setId(uniqueId);
        user.setDateJoined(new Date(System.currentTimeMillis()));
        user.setDateLastSeen(new Date(System.currentTimeMillis()));

        user.setPassword(encoder.encode(user.getPassword()));

        user.setRegistered(true); // todo add confirmations by email

        userRepository.save(user);
    }

    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        User user = username.contains("@") ? userRepository.findByEmail(username) : userRepository.findByUsername(username);
        if(user != null) {
            System.out.println("USER FROUND" + user.getUsername());
            List<GrantedAuthority> authorities = getUserAuthority(user.getGrants());
            return buildUserForAuthentication(user, authorities);
        } else {
            throw new UsernameNotFoundException("Username not found");
        }
    }

    private List<GrantedAuthority> getUserAuthority(List<Grant> grants) {
        List<GrantedAuthority> permissions = new ArrayList<>();
        for (Grant grant : grants) {
            if (grant.isActive()) {
                Rank rank = grant.getRank();
                rank.getCompoundedPermissions().forEach(node -> {
                    permissions.add(new SimpleGrantedAuthority(node));
                });
            }
        }

        return permissions;
    }

    private UserDetails buildUserForAuthentication(User user, List<GrantedAuthority> authorities) {
        return new org.springframework.security.core.userdetails.User(user.getUsername(), user.getPassword(), authorities);
    }

}

File: src/main/kotlin/dev/foraged/forums/user/service/UserService.kt
package dev.foraged.forums.user.service

import dev.foraged.forums.user.User
import dev.foraged.forums.user.UserRepository
import gg.scala.cache.uuid.ScalaStoreUuidCache
import gg.scala.store.storage.type.DataStoreStorageType
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.stereotype.Service
import java.util.*
import java.util.concurrent.CompletableFuture

@Service
class UserService : UserDetailsService
{
    val userRepository by lazy {
        UserRepository
    }
    @Autowired lateinit var encoder: BCryptPasswordEncoder

    fun findUserByEmail(email: String): User? {
        return userRepository.findByEmail(email)
    }

    fun findUserByRegisterSecret(email: String): User? {
        return userRepository.findByRegisterSecret(email)
    }

    fun findUserByName(name: String): User? {
        return userRepository.findByUsernameIgnoreCase(name)
    }

    fun findUserByUniqueId(uuid: UUID): User? {
        return userRepository.findByIdentifier(uuid)
    }

    fun save(user: User): CompletableFuture<Void> {
        return userRepository.controller.save(user, DataStoreStorageType.MONGO)
    }

    /**
     * Creating registered users for the first time
     *
     * @param user User to register
     */
    @Throws(Exception::class)
    fun createUser(user: User)
    {
        val uniqueId = ScalaStoreUuidCache.uniqueId(user.username) ?: throw RuntimeException("Error finding uniqueId")
        user.identifier = uniqueId
        user.password = encoder.encode(user.password)
        userRepository.controller.save(user)
    }

    @Throws(UsernameNotFoundException::class)
    override fun loadUserByUsername(username: String): UserDetails
    {
        val user =
            if (username.contains("@")) userRepository.findByEmail(username) else userRepository.findByUsername(
                username
            )
        return if (user != null)
        {
            println("USER FROUND" + user.username)
            val authorities = getUserAuthority(user)
            buildUserForAuthentication(user, authorities)
        } else
        {
            throw UsernameNotFoundException("Username not found")
        }
    }

    private fun getUserAuthority(user: User): List<GrantedAuthority>
    {
        val permissions: MutableList<GrantedAuthority> = ArrayList()
        // make this use other one too
        /* for (grant in user.activeGrants.filterNotNull()) grant.rank.compoundedPermissions.forEach {
            permissions.add(SimpleGrantedAuthority(it))
        }*/
        return permissions
    }

    private fun buildUserForAuthentication(user: User, authorities: List<GrantedAuthority>): UserDetails {
        return org.springframework.security.core.userdetails.User(user.username, user.password, authorities)
    }
}